Handle TriggerBuild matching multiple components

marmarek opened the pull request:
With self.config.get_components(..., url_match=True), there may be multiple
components to build. Add loop over components list where necessary. This
also allows not hardcoding `components[0]` in multiple places.

This is necessary to build linux-kernel and/or linux-kernel-latest when 
getting a push notification for linux-kernel repo.

marmarek committed:
Handle TriggerBuild matching multiple components

With self.config.get_components(..., url_match=True), there may be
multiple components to build. Add loop over components list where
necessary. This also allows not hardcoding `components[0]` in multiple
places.

This is necessary to build linux-kernel and/or linux-kernel-latest when
getting a push notification for linux-kernel repo.

marmarek committed:
tests: check --no-signer-github-command-check too

marmarek committed:
test_action: assert that rpm packages are not uploaded to security-testing

The upload command specifies vm-bullseye only, so test if that's really
the case. In fact, I'm not sure why the test passed before...

And modify test_rpc.py to upload to security-testing for Debian only
too.

marmarek committed:
Adjust handling non-existing templates, and fix getting build timestamp

Non-existing templates result now in ConfigError exception, instead of
being silently ignored.

marmarek committed:
Properly report build/upload stage failure

When upload fails as part of build command, report it as upload failure
too, not generic build failure.

The difference is that, failed build will be retried on next push, but
failied upload will not (but will be re-uploaded on subsequent build of
possibly other component).

marmarek committed:
test: do not swallow output

Let pytest collect it, and display in case of failure.

marmarek committed:
test: include local variables in failed results

marmarek committed:
ci: add mypy

marmarek committed:
Adjust for issues reported by mypy

- add few asserts (following args.command branches)
- don't reuse the same variable as different types
- define common abstract methods in the base class
- add missing 'manager' parameter to _upload() call

marmarek commented:
@fepitre this is ready for review

fepitre commented on github-action.py:
```diff
@@ -429,11 +440,12 @@ def build(self):
-        actual_commit_sha = self.components[0].get_source_commit_hash()
+        actual_commit_sha = self.component.get_source_commit_hash()
         if self.commit_sha != actual_commit_sha:
-            raise AutoActionError(
```
This is to prevent full stop for all dists or all matching components?

marmarek replied:
Yes, exactly. When component is matched using git repo name too, then one name can match multiple components (for example linux-kernel and linux-kernel-latest). The commit id will match only to one (or some) of them, those not matching should be skipped instead of preventing matching from being processed.

marmarek replied:
Hmm, in fact in the most recent version, I changed how multiple components are processed - now separate `cli` instances are used. So, if you prefer, I can keep exception here, but catch it in `for cli in cli_list` loop to allow other components being processed. Any preference?

fepitre replied:
Somehow I like raising exception to be sure we can manage it in the same place where we would catch all them. It may allow reusing the message for another thing like we would send a status or something else to report it.

marmarek replied:
Changed.

fepitre reviewed (approved):
It looks good.

marmarek commented:
PipelineRetry

marmarek committed:
Adjust for Config object changes

There is no Config.get_artifacts_dir() function anymore, it's a
Config.artifacts_dir property now.
